### PR TITLE
ospp: Feature/traffic

### LINKED ISSUE #402: GSoC 2022: Pixiu Traffic Distribution Mode Implementation
The gateway should provides the ability to distribute traffic, which makes it easier to build service o&M capabilities such as Canary testing and blue and green publishing.

So if you are interested in this project and the objective described in the proposal, please leave comments below so we can further exchange opinions.

Please don't forget to apply for GSoC from its official website when the application slot opens.

### PR DESCRIPTION
**Pull request by maxingg**
**What this PR does**:
Optimize the traffic filter, and make cluster config support both header and weight strategy.

**Which issue(s) this PR fixes**:

Fixes #402

### CODE REVIEW

File: pixiu/pkg/filter/traffic/traffic.go [hunk 1]
```diff
@@ -79,41 +91,88 @@ func (factory *FilterFactory) Config() interface{} {
...
-	initActions()
 	return nil
 }
 
 func (factory *FilterFactory) PrepareFilterChain(ctx *http.HttpContext, chain filter.FilterChain) error {
-	f := &Filter{}
-	f.Rules = factory.rulesMatch(ctx.Request.RequestURI)
+	f := &Filter{
+		weight: unInitialize,
+		record: map[string]bool{},
```
Comment (baerwang) [inline pixiu/pkg/filter/traffic/traffic.go]: bool types change struct{}

File: pixiu/pkg/filter/traffic/traffic.go [hunk 2]
```diff
@@ -79,41 +91,88 @@ func (factory *FilterFactory) Config() interface{} {
...
 
 func (f *Filter) Decode(ctx *http.HttpContext) filter.FilterStatus {
 	if f.Rules != nil {
-		cluster := spilt(ctx.Request, f.Rules)
-		if cluster != nil {
-			ctx.Route.Cluster = cluster.Name
-			logger.Debugf("[dubbo-go-pixiu] execute traffic split to cluster %s", cluster.Name)
+		for _, wp := range f.Rules {
+			result := f.traffic(wp, ctx)
+			if result {
```
Comment (baerwang) [inline pixiu/pkg/filter/traffic/traffic.go]:
```go
if f.traffic(wp, ctx) {}
```

File: pixiu/pkg/filter/traffic/traffic.go [hunk 3]
```diff
@@ -79,41 +91,88 @@ func (factory *FilterFactory) Config() interface{} {
...
+				}
+				if cluster.CanaryWeight != "" {
+					val, err := strconv.Atoi(cluster.CanaryWeight)
+					if err != nil || val <= 0 {
+						panic(fmt.Sprintf("Wrong canary-weight value: %v", cluster.CanaryWeight))
+					}
+					wp.weightFloor = up
+					up += val
+					if up > 100 {
+						panic("Total canary-weight more than 100")
```
Comment (baerwang) [inline pixiu/pkg/filter/traffic/traffic.go]: should not panic

File: pixiu/pkg/filter/traffic/traffic.go [hunk 4]
```diff
@@ -79,41 +91,88 @@ func (factory *FilterFactory) Config() interface{} {
...
+func (f *Filter) traffic(c *ClusterWrapper, ctx *http.HttpContext) bool {
+	if f.weight == unInitialize {
+		rand.Seed(time.Now().UnixNano())
+		f.weight = rand.Intn(100) + 1
+	}
+
+	res := false
+	if c.header != "" {
+		res = headerSpilt(ctx.Request, c.header)
+	} else if res == false && c.weightFloor != -1 && c.weightCeil != -1 {
```
Comment (baerwang) [inline pixiu/pkg/filter/traffic/traffic.go]: pls fix reviewdog

File: pixiu/pkg/filter/traffic/traffic.go (hunk 1 above)
Comment (maxingg) [reply pixiu/pkg/filter/traffic/traffic.go]: ok

File: pixiu/pkg/filter/traffic/traffic.go (hunk 2 above)
Comment (maxingg) [reply pixiu/pkg/filter/traffic/traffic.go]: ok

File: pixiu/pkg/filter/traffic/traffic.go (hunk 3 above)
Comment (maxingg) [reply pixiu/pkg/filter/traffic/traffic.go]: I' ll handle with log.Errorf

File: pixiu/pkg/filter/traffic/traffic.go (hunk 4 above)
Comment (maxingg) [reply pixiu/pkg/filter/traffic/traffic.go]: Okay

**Review by baerwang — APPROVED** [review]

File: pixiu/pkg/filter/traffic/models.go [hunk 5]
```diff
@@ -18,61 +18,13 @@
...
-			rand.Seed(time.Now().UnixNano())
-			num := rand.Intn(100) + 1
-			intValue, _ := strconv.Atoi(value)
-			return num <= intValue
-		},
-	}
-
-	actions.wrappers = append(actions.wrappers, headerAction)
-	actions.wrappers = append(actions.wrappers, weightAction)
+func headerSpilt(req *http.Request, value string) bool {
```
Comment (AlexStocks) [inline pixiu/pkg/filter/traffic/models.go]: 改成 spitHeader

File: pixiu/pkg/filter/traffic/models.go [hunk 6]
```diff
@@ -18,61 +18,13 @@
...
-		// call action, decide to return cluster or not
-		if cluster, exist := rules[wrapper.name]; exist {
-			header, value := cluster.Canary, cluster.Value
-			if wrapper.action(header, value, request) {
-				return cluster
-			}
-		}
-	}
-	return nil
+func weightSpilt(weight, floor, ceil int) bool {
```
Comment (AlexStocks) [inline pixiu/pkg/filter/traffic/models.go]: splitWeight

**Review by mark4z — APPROVED** [review]

**Review by MasterKenway — APPROVED** [review]

**Pull request merged**